### PR TITLE
Change: Pyblish Wait for pool futures to finish

### PR DESCRIPTION
**Pull request by Tilix4**
## Changelog Description
Wait for publish pool to finish before closing pyblish.

## Testing notes:
1. Regular publish

### CODE REVIEW

File: openpype/hosts/blender/plugins/publish/integrate_blend.py [hunk 1]
```diff
@@ -51,6 +51,7 @@ def process(self, instance):
         representations.update(instance.data.get("hero_representations", {}))
 
         # Run commands for all published representations
+        instance.data.setdefault("futures", [])
```
Comment (Sharkitty) [inline openpype/hosts/blender/plugins/publish/integrate_blend.py]: I don't find `futures` to be a very explicit name, or I don't understand the logic behind it.

File: openpype/hosts/blender/plugins/publish/integrate_blend.py [hunk 2]
```diff
@@ -115,5 +116,8 @@ def callback(id: ObjectId, future: Future):
                 f = pool.submit(subprocess.check_call, main_command, shell=False)
                 f.add_done_callback(partial(callback, repre_id))
 
+                # Keep future for waiting for it to finish at unpause
+                instance.data["futures"].append(f)
```
Comment (Sharkitty) [inline openpype/hosts/blender/plugins/publish/integrate_blend.py]: Same here

File: openpype/hosts/blender/plugins/publish/unpause_sync_server.py [hunk 3]
```diff
@@ -15,3 +16,10 @@ def process(self, context):
         manager = ModulesManager()
         sync_server_module = manager.modules_by_name["sync_server"]
         sync_server_module.unpause_server()
+
+        # Wait for all started futures to finish
+        for instance in context:
+            for future in instance.data.get("futures", []):
+                while future.running():
```
Comment (Sharkitty) [inline openpype/hosts/blender/plugins/publish/unpause_sync_server.py]: And same here.

File: openpype/hosts/blender/plugins/publish/integrate_blend.py (hunk 1 above)
Comment (Tilix4) [reply openpype/hosts/blender/plugins/publish/integrate_blend.py]: Actually `future` is the official name of what is returned by `pool.submit`: https://docs.python.org/3/library/asyncio-future.html

Comment (Tilix4) [reply openpype/hosts/blender/plugins/publish/integrate_blend.py]: But I understand it's not very explicit. Would "representations_processes" or "repre_processes" be better?

File: openpype/hosts/blender/plugins/publish/unpause_sync_server.py (hunk 3 above)
Comment (Tilix4) [reply openpype/hosts/blender/plugins/publish/unpause_sync_server.py]: But for this one, I think keeping the future name helps to understand what it is about.

File: openpype/hosts/blender/plugins/publish/integrate_blend.py (hunk 1 above)
Comment (Sharkitty) [reply openpype/hosts/blender/plugins/publish/integrate_blend.py]: I was not familiar with it, if I understand correctly, it's the expected result of an async task, right? In that case the `future` mention makes sense, but how about `async_futures`, to explicit that this is related to asynchronous programming?

File: openpype/hosts/blender/plugins/publish/unpause_sync_server.py (hunk 3 above)
Comment (Sharkitty) [reply openpype/hosts/blender/plugins/publish/unpause_sync_server.py]: Agreed

File: openpype/hosts/blender/plugins/publish/integrate_blend.py (hunk 1 above)
Comment (Tilix4) [reply openpype/hosts/blender/plugins/publish/integrate_blend.py]: future cannot be anything else than from async. Maybe `representations_futures`?

Comment (Sharkitty) [reply openpype/hosts/blender/plugins/publish/integrate_blend.py]: Okay, fine by me ^^

File: openpype/hosts/blender/plugins/publish/integrate_blend.py [hunk 4]
```diff
@@ -115,5 +116,8 @@ def callback(id: ObjectId, future: Future):
                 f = pool.submit(subprocess.check_call, main_command, shell=False)
                 f.add_done_callback(partial(callback, repre_id))
 
+                # Keep future for waiting for it to finish at unpause
+                instance.data["representations_futures"].append(f)
+
```
Comment (Sharkitty) [inline openpype/hosts/blender/plugins/publish/integrate_blend.py]: W293 blank line contains white space

**Comment by Sharkitty** [comment]
When testing on my machine (Linux - Pop!_OS 22.04) with the sync queue open next to my publish, I still had blank remote lines when publish finished. I don't know if it's just the time the UI updates itself, in which case it should be fine.

**Comment by Tilix4** [comment]
> When testing on my machine (Linux - Pop!_OS 22.04) with the sync queue open next to my publish, I still had blank remote lines when publish finished. I don't know if it's just the time the UI updates itself, in which case it should be fine.

The important information is the sync status. If it is `Paused` the issue remains. If `Queued` it's okay, the blank line is they are waiting for the next sync loop.

**Comment by Sharkitty** [comment]
> The important information is the sync status. If it is Paused the issue remains. If Queued it's okay, the blank line is they are waiting for the next sync loop.

I'll perform another test on windows to be sure.

**Review by Sharkitty — APPROVED** [review]
LGTM!

**Pull request merged**